src/storage/src/source/postgres.rs

```diff
@@ -335,37 +338,36 @@ impl SourceReader for PostgresSourceReader {
                 lsn,
                 end,
             })) => {
+                self.last_lsn = lsn;
                 if end {
-                    Ok(NextMessage::Ready(SourceMessageType::Finalized(
-                        SourceMessage {
-                            output,
-                            partition: PartitionId::None,
-                            offset: lsn.into(),
-                            upstream_time_millis: None,
-                            key: (),
-                            value,
-                            headers: None,
-                            specific_diff: diff,
-                        },
-                    )))
+                    let msg = SourceMessage {
+                        output,
+                        upstream_time_millis: None,
+                        key: (),
+                        value,
+                        headers: None,
+                    };
+                    let ts = (PartitionId::None, lsn.into());
+                    NextMessage::Ready(SourceMessageType::Finalized(Ok(msg), ts, diff))
                 } else {
-                    Ok(NextMessage::Ready(SourceMessageType::InProgress(
-                        SourceMessage {
-                            output,
-                            partition: PartitionId::None,
-                            offset: lsn.into(),
-                            upstream_time_millis: None,
-                            key: (),
-                            value,
-                            headers: None,
-                            specific_diff: diff,
-                        },
-                    )))
+                    let msg = SourceMessage {
+                        output,
+                        upstream_time_millis: None,
+                        key: (),
+                        value,
+                        headers: None,
+                    };
+                    let ts = (PartitionId::None, lsn.into());
+                    NextMessage::Ready(SourceMessageType::InProgress(Ok(msg), ts, diff))
                 }
             }
-            Some(Some(InternalMessage::Err(e))) => Err(e),
-            None => Ok(NextMessage::Pending),
-            Some(None) => Ok(NextMessage::Finished),
+            Some(Some(InternalMessage::Err(err))) => {
+                // XXX(petrosagg): we are fabricating a timestamp here!!
```

Comment:
It would be possible to push this down quite a bit. The ideal would be to add a new variant to the `Definite` type which includes the timestamp.

I think it might be worth taking a little time to try this: in theory this source already distinguishes between definite and indefinite errors. (Totally possible some are misclassified ofc!) Being able to capture the LSN for the definite errors here would demonstrate that the new approach is working well.

Reply:
It's a bit tricky to do that because a lot of definite errors are generated by a `try_definite!` call which would all have to be annotated with an LSN but some of them don't have a clear LSN to be annotated with. For example we do `try_definite!(self.validate_tables(publication_tables));`. What should the LSN be for that error if it happens?